src/utilities/update.ts

```diff
@@ -39,69 +39,145 @@ function projectValuesOntoExistingObj(dst: Record<string, unknown>, src: Record<
+      } else if (quotePushCondition || stringPushCondition || specialPushCondition ||
+        delimiterPushCondition || commentBeginCondition) {
+        if (commentBeginCondition) {
+          stack.push('=begin')
+        } else {
+          stack.push(char)
+        }
+        rangeStack.push([i])
+      }
+      j++
+    }
+  }
+  return ranges
+}
+
+function getDistinctRanges(ranges: number[][]): number[][] {
+  const output: number[][] = []
+  for (const [x, y] of ranges) {
+    if (x !== y) {
+      output.push([x, y])
     }
   }
+  return output
+}
 
-  return indexVal
+function joinStringsFromRanges(text: string, ranges: number[][]): string[] {
+  const lines = text.split('\n')
+  const output: string[] = []
+  let i = 0
+  while (i < lines.length) {
+    let found = false
+    for (const [x, y] of ranges) {
+      if (i >= x && i <= y) {
+        output.push(lines.slice(x, y + 1).join('\n'))
+        found = true
+        i = y
+        break
+      }
+    }
+    if (!found) {
+      output.push(lines[i])
+    }
+    i++
+  }
+  return output
 }
 
 /*
   This is the most likely thing to break if you are getting code formatting issues.
   Extract the existing describe blocks (what is actually run by inspec for validation)
 */
 export function getExistingDescribeFromControl(control: Control): string {
-  // Algorithm:
-  //   Locate the start and end of the control string
-  //   Update the end of the control that contains information (if empty lines are at the end of the control)
-  //   loop: until the start index is changed (loop is done from the bottom up)
-  //     Clean testing array entry line (removes any non-print characters)
-  //     if: line starts with meta-information 'tag' or 'ref'
-  //       set start index to found location
-  //       break out of the loop
-  //     end
-  //   end
-  //   Remove any empty lines after the start index (in any)
-  //   Extract the describe block from the audit control given the start and end indices
-  // Assumptions: 
-  //  1 - The meta-information 'tag' or 'ref' precedes the describe block
-  // Pros: Solves the potential issue with option 1, as the lookup for the meta-information
-  //       'tag' or 'ref' is expected to the at the beginning of the line.
   if (control.code) {
-    let existingDescribeBlock = ''
-    let indexStart = control.code.toLowerCase().indexOf('control')
-    let indexEnd = control.code.toLowerCase().trimEnd().lastIndexOf('end')
-    const auditControl = control.code.substring(indexStart, indexEnd).split('\n')
-
-    indexStart = 0
-    indexEnd = auditControl.length - 1
-    indexEnd = getIndexOfFirstLine(auditControl, indexEnd, '-')
-    let index = indexEnd
-
-    while (indexStart === 0) {
-      const line = auditControl[index].toLowerCase().trim()
-      if (line.indexOf('ref') === 0 || line.indexOf('tag') === 0) {
-        indexStart = index + 1
+    // Join multi-line strings in Control block.
+    let ranges = getRangesFromStackUpdate(control.code)
```

Comment:
Thinking of naming and clarity here. What do you mean by ranges and what might convey this simpler?